### ISSUE TITLE
stats logging in "on_train_epoch_end" ends up on wrong progress bar

### ISSUE THREAD

**Issue by jojje** [Lightning-AI/pytorch-lightning#19322]
### Bug description

When logging statistics at the end of an epoch from within `on_train_epoch_end`, the statistics end up on the wrong progress bar.

Since there doesn't seem to be a configuration to tell lightning nor the `TQDMProgressBar` to retain the bar for each epoch, I've been forced to inject a new line after each epoch ends, in order to not lose any of the valuable statistics in the console output.

The following is the output from a 3 epoch run:
```
Epoch 0: 100%|█████████████████████████| 938/938 [00:04<00:00, 206.06it/s, v_num=207]
Epoch 1: 100%|█████████████| 938/938 [00:04<00:00, 233.29it/s, v_num=207, loss=0.553]
Epoch 2: 100%|█████████████| 938/938 [00:04<00:00, 233.39it/s, v_num=207, loss=0.329]
`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████| 938/938 [00:04<00:00, 232.93it/s, v_num=207, loss=0.329]
```
* Loss for epoch 0 is incorrectly shown for epoch 1. 
* Loss for epoch 1 is incorrectly shown for epoch 2. 
* No logged loss at all is reported for epoch 0 nor epoch 2, 

If there is a _proper_ way to retain the progress bar for each epoch that is different from what I'm doing, then please let me know and this ticket can then be closed. If not, hopefully a fix can be found.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import torch
import torchvision
import pytorch_lightning as pl

class DemoNet(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(784, 10)
        self.batch_losses = []

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.001)

    def training_step(self, batch:torch.Tensor, _):
        x, y = batch
        x = x.reshape(x.size(0), -1)
        yh = self.fc(x)
        loss = torch.nn.functional.cross_entropy(yh, y)
        self.batch_losses.append(loss)
        return loss

    def on_train_epoch_end(self):
        loss = torch.stack(self.batch_losses).mean()
        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
        self.batch_losses.clear()
        print("")

ds = torchvision.datasets.MNIST(root="dataset/", train=True, transform=torchvision.transforms.ToTensor(), download=True)
train_loader = torch.utils.data.DataLoader(dataset=ds, batch_size=64, shuffle=False)
trainer = pl.Trainer(max_epochs=3)
trainer.fit(DemoNet(), train_loader)
```

### Error messages and logs

N/A

### Environment

<details>
  <summary>Current environment</summary>

* Lightning:
	- lightning-utilities: 0.9.0
	- pytorch-lightning: 2.1.3
	- torch:             2.1.2+cu118
	- torchaudio:        2.1.2+cu118
	- torchmetrics:      1.2.1
	- torchvision:       0.16.2+cu118
	- tqdm:              4.66.1
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:         AMD64 Family 25 Model 97 Stepping 2, AuthenticAMD
	- python:            3.10.0
	- release:           10
	- version:           10.0.19045
* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 4090
	- available:         True
	- version:           11.8

*  How you installed Lightning(`conda`, `pip`, source): pip
*  Running environment of LightningApp (e.g. local, cloud): local

</details>

cc @carmocca

**Comment by ManoBharathi93** [comment]
Injecting new line :
   ```
def on_train_epoch_end(self):
        loss = torch.stack(self.batch_losses).mean()
        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
        print(f"Epoch {self.current_epoch} --> loss{loss.item()}")
        self.batch_losses.clear()
        print("")
```
Output will look like this: 
![image](https://github.com/Lightning-AI/pytorch-lightning/assets/88357044/adca5152-6d90-4ef1-965c-5dfd9b857035)

@jojje  Do you meant this approach ?

cc @Borda Do you think it as issue or not? If issue can you please share your insights I could work on it.

**Comment by awaelchli** [comment]
Hey @jojje It could be seen as an issue or not, it depends. 
Fixing this might be very hard. It has to do with the fact that the callback hooks run before the LightningModule hooks if I interpret this correctly.

If you `self.log` in your training step with `on_epoch=True`, it will work correctly.

Regarding "why does Epoch 2 show twice" it is because you have print statements and the TQDM bar will continue to write updates to the progress bar after your prints. If you want to avoid that, use `self.print(...)` instead.

**Comment by jojje** [comment]
@awaelchli I tried the two changes you proposed, but it didn't solve the issue. Perhaps I misunderstood what you was trying to articulate.

Change:
```diff
@@ -18,11 +18,11 @@ class DemoNet(pl.LightningModule):
         loss = torch.nn.functional.cross_entropy(yh, y)
         self.batch_losses.append(loss)
+        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
         return loss

     def on_train_epoch_end(self):
         loss = torch.stack(self.batch_losses).mean()
-        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
         self.batch_losses.clear()
-        print("")
+        self.print("")

 ds = torchvision.datasets.MNIST(root="dataset/", train=True, transform=torchvision.transforms.ToTensor(), download=True)
```

Resulting output:
```


Epoch 2: 100%|██████████| 938/938 [00:03<00:00, 234.91it/s, v_num=5, loss=0.300]`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|██████████| 938/938 [00:03<00:00, 234.62it/s, v_num=5, loss=0.300]
```

As you can see, the previous progress bars are closed and thus not retained. That's why there are two leading blank lines. Additionally, there is still a final duplicate of the final progress bar for the third epoch (Epoch 2).

**Comment by jojje** [comment]
Update, workaround that makes lightning log as expected:

```
import torch
import torchvision
import pytorch_lightning as pl
from pytorch_lightning.callbacks.progress.tqdm_progress import TQDMProgressBar

class LitProgressBar(TQDMProgressBar):
    def on_train_end(self, *_):
        # self.train_progress_bar.close()
        pass

    def on_validation_end(self, trainer, pl_module):
        # self.val_progress_bar.close()
        self.reset_dataloader_idx_tracker()
        if self._train_progress_bar is not None and trainer.state.fn == "fit":
            self.train_progress_bar.set_postfix(self.get_metrics(trainer, pl_module))

    def on_test_end(self, trainer, pl_module):
        # self.test_progress_bar.close()
        self.reset_dataloader_idx_tracker()

    def on_predict_end(self, trainer, pl_module):
        # self.predict_progress_bar.close()
        self.reset_dataloader_idx_tracker()


class DemoNet(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(784, 10)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.001)

    def training_step(self, batch:torch.Tensor, _):
        x, y = batch
        x = x.reshape(x.size(0), -1)
        yh = self.fc(x)
        loss = torch.nn.functional.cross_entropy(yh, y)
        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
        return loss

    def on_train_epoch_end(self):
        # import ipdb; ipdb.set_trace(context=15)
        print("")
        pass

ds = torchvision.datasets.MNIST(root="dataset/", train=True, transform=torchvision.transforms.ToTensor(), download=True)
train_loader = torch.utils.data.DataLoader(dataset=ds, batch_size=64, shuffle=False)
trainer = pl.Trainer(max_epochs=3, callbacks=[LitProgressBar()])
trainer.fit(DemoNet(), train_loader)
```

The key bit of information here is the need to subclass the `TQDMProgressBar`, just to be able to disable all the hard-coded *`bar.close()` calls you make in the default progress bar.

It would be great if every user didn't have to deal with all that boiler plate for every project, and instead the TQDMProgressBar constructor taking an optional argument such as "leave:bool" (same as tqdm) that you'd then check in the code to decide whether to close the progress bars or not.

E.g. 

```
class TQDMProgressBar(ProgressBar):
    def __init__(self, refresh_rate: int = 1, process_position: int = 0, leave: bool = False):
```

**Comment by jojje** [comment]
[A PR](https://github.com/Lightning-AI/pytorch-lightning/pull/19578) for discussion and review has been submitted to address this issue.
If anyone has time to look at it and provide feedback, that'd be great.

_Reviewer note: There was a failed test, but it seems entirely unrelated. In fact, the change was made such that there is zero change in behavior by default, and explicitly setting a new flag (which no existing tests could possibly be aware of) is required to enable the new behavior , so I don't how this change could possibly be related to the failure of [core/test_metric_result_integration.py::test_result_reduce_ddp](https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=193051&view=logs&j=47e66f3c-897a-5428-da11-bf5c7745762e&t=b14bb263-6f50-5509-457c-7b46bb798cd1)_
